File: GameManager.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____Enums = require("Enums")
local GameStates = ____Enums.GameStates
local CharacterTypes = ____Enums.CharacterTypes
local ____MainMenu = require("Screens.MainMenu")
local MainMenu = ____MainMenu.default
local ____NewGameMenu = require("Screens.NewGameMenu")
local NewGameMenu = ____NewGameMenu.default
local ____PauseMenu = require("Screens.PauseMenu")
local PauseMenu = ____PauseMenu.default
local ____Board = require("Screens.Board")
local Board = ____Board.default
local ____WinScreen = require("Screens.WinScreen")
local WinScreen = ____WinScreen.default
local ____LoseScreen = require("Screens.LoseScreen")
local LoseScreen = ____LoseScreen.default
local ____Settings = require("Settings")
local Settings = ____Settings.default
local ____Helpers = require("Helpers")
local exhaustiveGuard = ____Helpers.exhaustiveGuard
local isEmpty = ____Helpers.isEmpty
local GameStateManager = require("Libraries.GameStateManager-main.gamestateManager")
local ____Player = require("Player")
local Player = ____Player.default
local ____Map = require("Screens.Map.Map")
local Map = ____Map.default
local ____Enemy = require("Enemies.Enemy")
local Enemy = ____Enemy.default
local ____Shop = require("Screens.Shop")
local Shop = ____Shop.default
local ____LevelUpScreen = require("Screens.LevelUpScreen")
local LevelUpScreen = ____LevelUpScreen.default
local ____PerkScreen = require("Screens.PerkScreen")
local PerkScreen = ____PerkScreen.default
local push = require("Libraries.push")
local ____AssetManager = require("Assets.AssetManager")
local AssetManager = ____AssetManager.default
local ____TextManager = require("Assets.TextManager")
local TextManager = ____TextManager.default
local ____AnimationManager = require("Assets.Animations.AnimationManager")
local AnimationManager = ____AnimationManager.default
local ____MusicPlayer = require("Assets.Music.MusicPlayer")
local MusicPlayer = ____MusicPlayer.default
local ____Grass = require("Biomes.Grass")
local Grass = ____Grass.default
____exports.default = __TS__Class()
local GameManager = ____exports.default
GameManager.name = "GameManager"
function GameManager.prototype.____constructor(self)
    self.devMode = false
    self.gameState = GameStates.MAIN_MENU
    self.player = __TS__New(Player, self)
    self.settings = __TS__New(Settings)
    self.map = __TS__New(Map, self)
    self.assetManager = __TS__New(AssetManager, self)
    self.animationManager = __TS__New(AnimationManager, self)
    self.musicPlayer = __TS__New(MusicPlayer, self)
    self.biome = __TS__New(Grass)
    if not self.devMode then
        TextManager:setDefaultFont()
    end
end
function GameManager.prototype.getCharacter(self, characterType)
    repeat
        local ____switch5 = characterType
        local ____cond5 = ____switch5 == CharacterTypes.PLAYER
        if ____cond5 then
            return self.player
        end
        ____cond5 = ____cond5 or ____switch5 == CharacterTypes.ENEMY
        if ____cond5 then
            local ____opt_0 = self.board
            return ____opt_0 and ____opt_0.enemy
        end
    until true
end
function GameManager.prototype.switchBasedOnGameState(self, gameState, enemy)
    if gameState == nil then
        gameState = self.gameState
    end
    if enemy == nil then
        local ____opt_2 = self.board
        enemy = ____opt_2 and ____opt_2.enemy
    end
    self.assetManager = __TS__New(AssetManager, self)
    self.musicPlayer:play(gameState, self.biome)
    repeat
        local ____switch7 = gameState
        local ____cond7 = ____switch7 == GameStates.MAIN_MENU
        if ____cond7 then
            self:switchToMainMenu()
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.NEW_GAME_MENU
        if ____cond7 then
            self:switchToNewGameMenu()
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.BOARD
        if ____cond7 then
            self:switchToBoard(enemy)
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.PAUSE_MENU
        if ____cond7 then
            self:switchToPauseMenu()
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.WIN_SCREEN
        if ____cond7 then
            self:switchToWinScreen()
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.LOSE_SCREEN
        if ____cond7 then
            self:switchToLoseScreen()
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.MAP
        if ____cond7 then
            self:switchToMap()
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.SHOP
        if ____cond7 then
            self:switchToShop()
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.LEVEL_UP
        if ____cond7 then
            self:switchToLevelUpScreen()
            break
        end
        ____cond7 = ____cond7 or ____switch7 == GameStates.PERKS
        if ____cond7 then
            self:switchToPerkScreen()
            break
        end
        do
            exhaustiveGuard(gameState)
        end
    until true
end
function GameManager.prototype.switchToMainMenu(self)
    local mainMenuState = {update = function(____, dt)
        local ____opt_4 = self.mainMenu
        if ____opt_4 ~= nil then
            ____opt_4:drawScreen()
        end
    end}
    self.gameState = GameStates.MAIN_MENU
    self.board = nil
    self.winScreen = nil
    self.loseScreen = nil
    self.shop = nil
    self.levelUpScreen = nil
    self.perkScreen = nil
    if isEmpty(self.mainMenu) then
        self.mainMenu = __TS__New(MainMenu, self)
    end
    GameStateManager:setState(mainMenuState)
end
function GameManager.prototype.switchToNewGameMenu(self)
    local newGameMenuState = {update = function(____, dt)
        local ____opt_6 = self.newGameMenu
        if ____opt_6 ~= nil then
            ____opt_6:drawScreen()
        end
    end}
    self.gameState = GameStates.NEW_GAME_MENU
    if isEmpty(self.newGameMenu) then
        self.newGameMenu = __TS__New(NewGameMenu, self)
    end
    GameStateManager:setState(newGameMenuState)
end
function GameManager.prototype.switchToPauseMenu(self)
    local pauseMenuState = {update = function(____, dt)
        local ____opt_8 = self.pauseMenu
        if ____opt_8 ~= nil then
            ____opt_8:drawScreen()
        end
    end}
    if isEmpty(self.pauseMenu) then
        self.pauseMenu = __TS__New(PauseMenu, self)
    end
    GameStateManager:setState(pauseMenuState)
end
function GameManager.prototype.switchToBoard(self, enemy)
    local boardState = {
        update = function(____, dt)
            local ____opt_10 = self.board
            if ____opt_10 ~= nil then
                ____opt_10:drawBoard()
            end
            self.assetManager:handleMouseHover()
            self.animationManager:updateAnimations(dt)
        end,
        draw = function()
            if not self.devMode then
                push:start()
                self.assetManager:drawAssets()
                push:finish()
            end
        end,
        mousepressed = function(____, x, y, button)
            if not self.devMode then
                self.assetManager:handleMousePressed(x, y, button)
            end
        end,
        mousereleased = function(____, x, y, button)
            if not self.devMode then
                self.assetManager:handleMouseReleased(x, y, button)
            end
        end
    }
    self.gameState = GameStates.BOARD
    self.winScreen = nil
    self.loseScreen = nil
    self.shop = nil
    self.levelUpScreen = nil
    self.perkScreen = nil
    if isEmpty(self.board) then
        self.board = __TS__New(
            Board,
            self,
            enemy or __TS__New(Enemy, self)
        )
        self.board.dealer:setup()
    end
    self.board:buildAssets()
    GameStateManager:setState(boardState)
end
function GameManager.prototype.switchToWinScreen(self)
    local winState = {update = function(____, dt)
        local ____opt_12 = self.winScreen
        if ____opt_12 ~= nil then
            ____opt_12:drawScreen()
        end
    end}
    self.gameState = GameStates.WIN_SCREEN
    self.loseScreen = nil
    self.shop = nil
    self.levelUpScreen = nil
    self.perkScreen = nil
    if isEmpty(self.winScreen) then
        self.winScreen = __TS__New(WinScreen, self)
    end
    GameStateManager:setState(winState)
end
function GameManager.prototype.switchToLoseScreen(self)
    local loseState = {update = function(____, dt)
        local ____opt_14 = self.loseScreen
        if ____opt_14 ~= nil then
            ____opt_14:drawScreen()
        end
    end}
    self.gameState = GameStates.LOSE_SCREEN
    self.board = nil
    self.winScreen = nil
    self.shop = nil
    self.levelUpScreen = nil
    self.perkScreen = nil
    if isEmpty(self.loseScreen) then
        self.loseScreen = __TS__New(LoseScreen)
    end
    GameStateManager:setState(loseState)
end
function GameManager.prototype.switchToMap(self)
    local mapState = {
        update = function(____, dt)
            self.map:drawMap()
        end,
        draw = function()
            self.map:drawBackground()
        end
    }
    self.gameState = GameStates.MAP
    self.board = nil
    self.winScreen = nil
    self.loseScreen = nil
    self.shop = nil
    self.levelUpScreen = nil
    self.perkScreen = nil
    GameStateManager:setState(mapState)
end
function GameManager.prototype.switchToShop(self)
    local shopState = {update = function(____, dt)
        local ____opt_16 = self.shop
        if ____opt_16 ~= nil then
            ____opt_16:drawShop()
        end
    end}
    self.gameState = GameStates.SHOP
    self.board = nil
    self.winScreen = nil
    self.loseScreen = nil
    self.levelUpScreen = nil
    self.perkScreen = nil
    if isEmpty(self.shop) then
        self.shop = __TS__New(Shop, self)
        self.shop:setup()
    end
    GameStateManager:setState(shopState)
end
function GameManager.prototype.switchToLevelUpScreen(self)
    local levelUpState = {update = function(____, dt)
        local ____opt_18 = self.levelUpScreen
        if ____opt_18 ~= nil then
            ____opt_18:drawScreen()
        end
    end}
    self.gameState = GameStates.LEVEL_UP
    self.board = nil
    self.winScreen = nil
    self.loseScreen = nil
    self.shop = nil
    self.perkScreen = nil
    if isEmpty(self.levelUpScreen) then
        self.levelUpScreen = __TS__New(LevelUpScreen, self)
        self.levelUpScreen:setup()
    end
    GameStateManager:setState(levelUpState)
end
function GameManager.prototype.switchToPerkScreen(self)
    local perkState = {update = function(____, dt)
        local ____opt_20 = self.perkScreen
        if ____opt_20 ~= nil then
            ____opt_20:drawScreen()
        end
    end}
    self.gameState = GameStates.PERKS
    self.board = nil
    self.winScreen = nil
    self.loseScreen = nil
    self.shop = nil
    self.levelUpScreen = nil
    if isEmpty(self.perkScreen) then
        self.perkScreen = __TS__New(PerkScreen, self)
    end
    GameStateManager:setState(perkState)
end
function GameManager.prototype.getCard(self, id)
    for ____, card in ipairs(self.player.hand) do
        if card.id == id then
            return card
        end
    end
    for ____, card in ipairs(self.player.deck) do
        if card.id == id then
            return card
        end
    end
    for ____, card in ipairs(self.player.discardPile) do
        if card.id == id then
            return card
        end
    end
    local ____opt_22 = self.board
    local enemy = ____opt_22 and ____opt_22.enemy
    if isEmpty(enemy) then
        return
    end
    for ____, card in ipairs(enemy.hand) do
        if card.id == id then
            return card
        end
    end
    for ____, card in ipairs(enemy.deck) do
        if card.id == id then
            return card
        end
    end
    for ____, card in ipairs(enemy.discardPile) do
        if card.id == id then
            return card
        end
    end
end
return ____exports

File: GameManager.ts
/** @noSelfInFile */

import { GameStates, CharacterTypes } from "./Enums";
import MainMenu from "Screens/MainMenu";
import NewGameMenu from "Screens/NewGameMenu";
import PauseMenu from "Screens/PauseMenu";
import Board from "Screens/Board";
import WinScreen from "Screens/WinScreen";
import LoseScreen from "Screens/LoseScreen";
import Settings from "Settings";
import Character from "Character";
import { exhaustiveGuard, isEmpty } from "Helpers";
import * as GameStateManager from "Libraries.GameStateManager-main.gamestateManager";
import Player from "Player";
import Map from "Screens/Map/Map";
import Enemy from "Enemies/Enemy";
import Shop from "Screens/Shop";
import LevelUpScreen from "Screens/LevelUpScreen";
import PerkScreen from "Screens/PerkScreen";
import * as push from "Libraries.push";
import AssetManager from "Assets/AssetManager";
import Card from "Cards/Card";
import AnimationManager from "Assets/Animations/AnimationManager";
import MusicPlayer from "Assets/Music/MusicPlayer";
import Biome from "Biomes/Biome";
import Grass from "Biomes/Grass";
import BackgroundManager from "Screens/BackgroundManager";
import ShaderManager from "Shaders/ShaderManager";

interface GameState {
  update: (dt: number) => void;
  draw?: () => void;
  mousepressed?: (x: number, y: number, button: number) => void;
  mousereleased?: (x: number, y: number, button: number) => void;
}

export default class GameManager {
  gameState: GameStates;
  player = new Player(this);
  settings = new Settings();
  mainMenu?: MainMenu;
  newGameMenu?: NewGameMenu;
  pauseMenu?: PauseMenu;
  board?: Board;
  winScreen?: WinScreen;
  loseScreen?: LoseScreen;
  map: Map = new Map(this);
  shop?: Shop;
  levelUpScreen?: LevelUpScreen;
  perkScreen?: PerkScreen;
  assetManager = new AssetManager(this);
  animationManager = new AnimationManager(this);
  musicPlayer = new MusicPlayer(this);
  biome: Biome
  backgroundManager = new BackgroundManager(this);
  shaderManager = new ShaderManager(this);
  devMode: boolean = false; // Change if you want to test in dev mode

  constructor() {
    this.gameState = GameStates.MAIN_MENU;
    this.biome = new Grass(); // TODO: initialize this based on data from the map
  }

  getCharacter(characterType: string): Character | undefined {
    switch (characterType) {
      case CharacterTypes.PLAYER:
        return this.player;
      case CharacterTypes.ENEMY:
        return this.board?.enemy;
    }
  }

  switchBasedOnGameState(gameState = this.gameState, enemy = this.board?.enemy): void {
    this.assetManager = new AssetManager(this);
    this.backgroundManager.updateBackground(gameState);
    if (this.settings.playMusic) {
      this.musicPlayer.play(gameState, this.biome);
    }

    switch (gameState) {
      case GameStates.MAIN_MENU:
        this.switchToMainMenu();
        break;
      case GameStates.NEW_GAME_MENU:
        this.switchToNewGameMenu();
        break;
      case GameStates.BOARD:
        this.switchToBoard(enemy);
        break;
      case GameStates.PAUSE_MENU:
        this.switchToPauseMenu();
        break;
      case GameStates.WIN_SCREEN:
        this.switchToWinScreen();
        break;
      case GameStates.LOSE_SCREEN:
        this.switchToLoseScreen();
        break;
      case GameStates.MAP:
        this.switchToMap();
        break;
      case GameStates.SHOP:
        this.switchToShop();
        break;
      case GameStates.LEVEL_UP:
        this.switchToLevelUpScreen();
        break;
      case GameStates.PERKS:
        this.switchToPerkScreen();
        break;
      default:
        exhaustiveGuard(gameState);
    }
  }

  private switchToMainMenu(): void {
    const mainMenuState: GameState = {
      update: (dt: number) => {
        this.mainMenu?.drawScreen();
      },
    };

    this.gameState = GameStates.MAIN_MENU;
    this.board = undefined;
    this.winScreen = undefined;
    this.loseScreen = undefined;
    this.shop = undefined;
    this.levelUpScreen = undefined;
    this.perkScreen = undefined;

    if (isEmpty(this.mainMenu)) {
      this.mainMenu = new MainMenu(this);
    }

    GameStateManager.setState(mainMenuState);
  }

  private switchToNewGameMenu(): void {
    const newGameMenuState: GameState = {
      update: (dt: number) => {
        this.newGameMenu?.drawScreen();
      },
    };

    this.gameState = GameStates.NEW_GAME_MENU;

    if (isEmpty(this.newGameMenu)) {
      this.newGameMenu = new NewGameMenu(this);
    }

    GameStateManager.setState(newGameMenuState);
  }

  private switchToPauseMenu(): void {
    const pauseMenuState: GameState = {
      update: (dt: number) => {
        this.pauseMenu?.drawScreen();
      },
    };

    // Game state needs to be the previous state in the pause menu so we save correctly
    // this.gameState = GameStates.PAUSE_MENU


    if (isEmpty(this.pauseMenu)) {
      this.pauseMenu = new PauseMenu(this);
    }

    GameStateManager.setState(pauseMenuState);
  }

  private switchToBoard(enemy?: Enemy): void {
    const shadowShader = love.graphics.newShader("Shaders/Shadow.glsl");
    shadowShader.send("shadow_strength", 0.6);
    shadowShader.send("shadow_color", [0, 0, 0]); 

    const boardState: GameState = {
      update: (dt: number) => {        
        this.assetManager.handleMouseHover();        
        this.animationManager.updateAnimations(dt);
        this.shaderManager.updateShaders(dt);
      },
      draw: () => {
        if (!this.devMode) {
          push.start();
          //love.graphics.setShader(shadowShader);
          this.assetManager.drawAssets();
          love.graphics.setShader();  // Reset shader to default
          push.finish();
        }
      },
      mousepressed: (x: number, y: number, button: number) => {
        if (!this.devMode) {
          this.assetManager.handleMousePressed(x, y, button);
        }
      },
      mousereleased: (x: number, y: number, button: number) => {
        if (!this.devMode) {
          this.assetManager.handleMouseReleased(x, y, button);
        }
      },
    };

    this.gameState = GameStates.BOARD;
    this.winScreen = undefined;
    this.loseScreen = undefined;
    this.shop = undefined;
    this.levelUpScreen = undefined;
    this.perkScreen = undefined;
    
    if (isEmpty(this.board)) {
      this.board = new Board(this, enemy ?? new Enemy(this));
      this.board.dealer.setup();
    }

    this.board.start();

    GameStateManager.setState(boardState);
  }

  private switchToWinScreen(): void {
    const winState: GameState = {
      update: (dt: number) => {
        this.winScreen?.drawScreen();
      },
    };

    this.gameState = GameStates.WIN_SCREEN;
    // We need some data from the board to show stats and loot cards, so we keep it
    // this.board = {}
    this.loseScreen = undefined;
    this.shop = undefined;
    this.levelUpScreen = undefined;
    this.perkScreen = undefined;

    if (isEmpty(this.winScreen)) {
      this.winScreen = new WinScreen(this);
    }

    GameStateManager.setState(winState);
  }

  private switchToLoseScreen(): void {
    const loseState: GameState = {
      update: (dt: number) => {
        this.loseScreen?.drawScreen();
      },
    };

    this.gameState = GameStates.LOSE_SCREEN;
    this.board = undefined;
    this.winScreen = undefined;
    this.shop = undefined;
    this.levelUpScreen = undefined;
    this.perkScreen = undefined;

    if (isEmpty(this.loseScreen)) {
      this.loseScreen = new LoseScreen();
    }

    GameStateManager.setState(loseState);
  }

  private switchToMap(): void {
    const mapState: GameState = {
      update: (dt: number) => {
        this.map.drawMap();
      },
      draw: () => {
        this.map.drawBackground();
      },
    };

    this.gameState = GameStates.MAP;

    this.board = undefined;
    this.winScreen = undefined;
    this.loseScreen = undefined;
    this.shop = undefined;
    this.levelUpScreen = undefined;
    this.perkScreen = undefined;

    GameStateManager.setState(mapState);
  }

  private switchToShop(): void {
    const shopState: GameState = {
      update: (dt: number) => {
        this.shop?.drawShop();
      },
    };

    this.gameState = GameStates.SHOP;

    this.board = undefined;
    this.winScreen = undefined;
    this.loseScreen = undefined;
    this.levelUpScreen = undefined;
    this.perkScreen = undefined;

    if (isEmpty(this.shop)) {
      this.shop = new Shop(this);
      this.shop.setup();
    }

    GameStateManager.setState(shopState);
  }

  private switchToLevelUpScreen(): void {
    const levelUpState: GameState = {
      update: (dt: number) => {
        this.levelUpScreen?.drawScreen();
      },
    };

    this.gameState = GameStates.LEVEL_UP;

    this.board = undefined;
    this.winScreen = undefined;
    this.loseScreen = undefined;
    this.shop = undefined;
    this.perkScreen = undefined;

    if (isEmpty(this.levelUpScreen)) {
      this.levelUpScreen = new LevelUpScreen(this);
      this.levelUpScreen.setup();
    }

    GameStateManager.setState(levelUpState);
  }

  private switchToPerkScreen(): void {
    const perkState: GameState = {
      update: (dt: number) => {
        this.perkScreen?.drawScreen();
      },
    };

    this.gameState = GameStates.PERKS;

    this.board = undefined;
    this.winScreen = undefined;
    this.loseScreen = undefined;
    this.shop = undefined;
    this.levelUpScreen = undefined;

    if (isEmpty(this.perkScreen)) {
      this.perkScreen = new PerkScreen(this);
    }

    GameStateManager.setState(perkState);
  }

  // TODO: this is, of course, an extended search that we could optimize if necessary
  getCard(id: string): Card | undefined {
    for (const card of this.player.hand) {
      if (card.id === id) {
        return card;
      }
    }

    for (const card of this.player.deck) {
      if (card.id === id) {
        return card;
      }
    }

    for (const card of this.player.discardPile) {
      if (card.id === id) {
        return card;
      }
    }

    const enemy = this.board?.enemy;
    if (isEmpty(enemy)) {
      return;
    }

    for (const card of enemy.hand) {
      if (card.id === id) {
        return card;
      }
    }

    for (const card of enemy.deck) {
      if (card.id === id) {
        return card;
      }
    }

    for (const card of enemy.discardPile) {
      if (card.id === id) {
        return card;
      }
    }
  }
}
